### ISSUE TITLE
Poor end of speech detection

### ISSUE THREAD

**Issue by HarvsG** [rhasspy/wyoming-satellite#20]
I am running the satellite with local wake word detection on a raspberry pi 3 with an omnidirectional microphone and no speaker.
Wake words are detected almost instantly (I see the Assist In progress sensor change quickly). However for the light to actually turn off takes about 10 seconds (Assist In progress stays on for all of this time).
It seems to be that it is not detecting the end of speech well.
```yaml
stage: done
run:
  pipeline: 01gzv00ms6vap6j6fa7cn6t2s4
  language: en
events:
  - type: run-start
    data:
      pipeline: 01gzv00ms6vap6j6fa7cn6t2s4
      language: en
    timestamp: "2023-12-19T15:47:31.366949+00:00"
  - type: stt-start
    data:
      engine: cloud
      metadata:
        language: en-GB
        format: wav
        codec: pcm
        bit_rate: 16
        sample_rate: 16000
        channel: 1
    timestamp: "2023-12-19T15:47:31.367373+00:00"
  - type: stt-vad-end
    data:
      timestamp: 7500
    timestamp: "2023-12-19T15:47:46.339065+00:00"
  - type: stt-end
    data:
      stt_output:
        text: Turn on the study lamp.
    timestamp: "2023-12-19T15:47:46.593086+00:00"
  - type: intent-start
    data:
      engine: homeassistant
      language: en
      intent_input: Turn on the study lamp.
      conversation_id: null
      device_id: 87fd30dbb98112109029d1a8689bba24
    timestamp: "2023-12-19T15:47:46.593309+00:00"
  - type: intent-end
    data:
      intent_output:
        response:
          speech:
            plain:
              speech: Turned on light
              extra_data: null
          card: {}
          language: en
          response_type: action_done
          data:
            targets: []
            success:
              - name: Study Lamp
                type: entity
                id: light.study_lamp
            failed: []
        conversation_id: null
    timestamp: "2023-12-19T15:47:46.784007+00:00"
  - type: run-end
    data: null
    timestamp: "2023-12-19T15:47:46.784109+00:00"
stt:
  engine: cloud
  metadata:
    language: en-GB
    format: wav
    codec: pcm
    bit_rate: 16
    sample_rate: 16000
    channel: 1
  done: true
  stt_output:
    text: Turn on the study lamp.
intent:
  engine: homeassistant
  language: en
  intent_input: Turn on the study lamp.
  conversation_id: null
  device_id: 87fd30dbb98112109029d1a8689bba24
  done: true
  intent_output:
    response:
      speech:
        plain:
          speech: Turned on light
          extra_data: null
      card: {}
      language: en
      response_type: action_done
      data:
        targets: []
        success:
          - name: Study Lamp
            type: entity
            id: light.study_lamp
        failed: []
    conversation_id: null
```

journalctl produces no live output
```
george@redpi:~ $ journalctl -u wyoming-satellite.service -f
-- Journal begins at Tue 2023-12-05 06:31:03 GMT. --
Dec 19 15:41:32 redpi systemd[1]: Started Wyoming Satellite.
Dec 19 15:41:33 redpi run[624]: INFO:root:Ready
Dec 19 15:41:33 redpi run[624]: INFO:root:Connected to services
Dec 19 15:41:33 redpi run[633]: Recording raw data 'stdin' : Signed 16 bit Little Endian, Rate 16000 Hz, Mono
Dec 19 15:42:35 redpi run[624]: INFO:root:Waiting for wake word
Dec 19 15:42:55 redpi run[624]: INFO:root:Waiting for wake word
Dec 19 15:43:22 redpi run[624]: INFO:root:Waiting for wake word
Dec 19 15:44:38 redpi run[624]: INFO:root:Waiting for wake word
Dec 19 15:47:27 redpi run[624]: INFO:root:Waiting for wake word
Dec 19 15:47:46 redpi run[624]: INFO:root:Waiting for wake word
Dec 19 15:48:44 redpi run[624]: INFO:root:Waiting for wake word
```
```
george@redpi:~ $ journalctl -u wyoming-openwakeword.service -f
Dec 19 15:41:24 redpi systemd[1]: Started Wyoming openWakeWord.
Dec 19 15:41:29 redpi run[519]: INFO:root:Ready
Dec 19 15:41:34 redpi run[519]: INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
```

```ini
[Unit]
Description=Wyoming Satellite
Wants=network-online.target
After=network-online.target
Requires=wyoming-openwakeword.service

[Service]
Type=simple
ExecStart=/home/george/wyoming-satellite/script/run --name 'Study Satellite' --uri 'tcp://0.0.0.0:10400' --mic-command 'arecord -D plughw:CARD=Device,DEV=0 -r 16000 -c 1 -f S16_LE -t raw' --wake-word-name 'hey_rhasspy'
WorkingDirectory=/home/george/wyoming-satellite
Restart=always
RestartSec=1

[Install]
WantedBy=default.target
```

```ini
[Unit]
Description=Wyoming openWakeWord

[Service]
Type=simple
ExecStart=/home/george/wyoming-openwakeword/script/run --uri 'tcp://127.0.0.1:10400'
WorkingDirectory=/home/george/wyoming-openwakeword
Restart=always
RestartSec=1

[Install]
WantedBy=default.target
```

**Comment by HarvsG** [comment]
Could it be some timeout occurring because I am not using a speaker? Although in that case I would have thought the light would still turn off/on quickly because the TTS should happen after the action?
P.S I ran this voice satellite after chapter 4 and had near-instantaneous responses

Whilst troubleshooting I ran into this error
```
Dec 19 16:16:14 redpi systemd[1]: Stopping Wyoming Satellite...
Dec 19 16:16:14 redpi systemd[1]: wyoming-satellite.service: Succeeded.
Dec 19 16:16:14 redpi systemd[1]: Stopped Wyoming Satellite.
Dec 19 16:16:14 redpi systemd[1]: wyoming-satellite.service: Consumed 1.333s CPU time.
Dec 19 16:16:14 redpi systemd[1]: Started Wyoming Satellite.
Dec 19 16:16:15 redpi run[779]: INFO:root:Ready
Dec 19 16:16:15 redpi run[779]: INFO:root:Connected to services
Dec 19 16:16:15 redpi run[787]: Recording raw data 'stdin' : Signed 16 bit Little Endian, Rate 16000 Hz, Mono
Dec 19 16:16:15 redpi run[787]: Aborted by signal Terminated...
Dec 19 16:16:15 redpi run[787]: arecord: pcm_read:2152: read error: Interrupted system call
```

**Comment by sebirdman** [comment]
@HarvsG i've seen similar issues, but only sometimes. It appears to be worse when some other sound nearby is playing (like a TV or some music)

I've had some success playing with the `--mic-noise-suppression` features documented here - https://github.com/rhasspy/wyoming-satellite?tab=readme-ov-file#audio-enhancements

but it's still a bit hit or miss.

**Comment by HarvsG** [comment]
> @HarvsG i've seen similar issues, but only sometimes. It appears to be worse when some other sound nearby is playing (like a TV or some music)
> 
> I've had some success playing with the `--mic-noise-suppression` features documented here - https://github.com/rhasspy/wyoming-satellite?tab=readme-ov-file#audio-enhancements
> 
> but it's still a bit hit or miss.

 --mic-noise-suppression 2 seemed to do very little but --mic-auto-gain 5 seems to reduce the wakeword hit rate from about 90% to about 70% but also reduce the 10 second delay problem to about 30% of the time.

in the meantime I have enabled recordings of the audio clips as per this: https://www.home-assistant.io/voice_control/troubleshooting/#to-tweak-the-assist-audio-configuration-for-your-device and will tweak settings.

**Comment by HarvsG** [comment]
Seen on the forum: https://community.home-assistant.io/t/home-assistant-cloud-stt-latency/657006

**Comment by synesthesiam** [comment]
I wonder if someone more knowledgeable about webrtc could tell me if I'm doing something wrong: https://github.com/rhasspy/webrtc-noise-gain/blob/32a122569ca3ac55177570c30626d27f6dd1fc1a/python.cpp#L44

The comments here suggest you can have both gain controllers active at the same time, but some of the examples are out of date: https://github.com/rhasspy/webrtc-noise-gain/blob/32a122569ca3ac55177570c30626d27f6dd1fc1a/webrtc-audio-processing/webrtc-audio-processing-1/modules/audio_processing/include/audio_processing.h#L157

In any case, it's unclear to me the exact interactions you can get when combining noise suppression, auto gain, and volume amplification. The VAD signal used for silence detection is also coming from here, so that makes it extra confusing because the auto gain uses VAD to change volume (AFAIK). I'd prefer to switch to [silero-vad](https://github.com/snakers4/silero-vad) in core not only because it's a better VAD, but because it's also not coupled with the other audio stuff.

**Comment by HarvsG** [comment]
> I wonder if someone more knowledgeable about webrtc could tell me if I'm doing something wrong: https://github.com/rhasspy/webrtc-noise-gain/blob/32a122569ca3ac55177570c30626d27f6dd1fc1a/python.cpp#L44
> 
> The comments here suggest you can have both gain controllers active at the same time, but some of the examples are out of date: https://github.com/rhasspy/webrtc-noise-gain/blob/32a122569ca3ac55177570c30626d27f6dd1fc1a/webrtc-audio-processing/webrtc-audio-processing-1/modules/audio_processing/include/audio_processing.h#L157
> 
> In any case, it's unclear to me the exact interactions you can get when combining noise suppression, auto gain, and volume amplification. The VAD signal used for silence detection is also coming from here, so that makes it extra confusing because the auto gain uses VAD to change volume (AFAIK). I'd prefer to switch to [silero-vad](https://github.com/snakers4/silero-vad) in core not only because it's a better VAD, but because it's also not coupled with the other audio stuff.

What part of the engine/pipeline is responsible for detecting the end of speech? Is it done by VAD on the satellite, or within the STT engine?

**Comment by smoltron** [comment]
I have this same problem that I posted here: https://community.home-assistant.io/t/home-assistant-cloud-stt-latency/657006
I get always a 15 second wait until the action is done.  There are no major problems in speech recognition. I do not have a satellite but a local microphone, Jabra 410, which is a very good and accurate one. My pipeline is composed of these:
Language: Finnish
STT: home assistant cloud
TTS: home assistant cloud
Wake word engine: openwakeword
Audio: Assist microphone

**Comment by vhsdream** [comment]
I want to chime in to say that I do not have this issue and my setup is as follows:

satellite HW: pi 3B
Language: en
Wake word engine: openwakeword (one in HA, one running on satellite)
STT: faster-whisper (in LXC with passed through GPU)
TTS: piper (in HA)
Audio: Anker S330

I don't know why, but I have a hunch that the issue is with Home Assistant Cloud. Can someone switch to whisper for STT, then test, and then switch to piper for TTS and test that, see if the issue still presents itself?

**Comment by smoltron** [comment]
I tried Whisper and Piper. Whisper seems not to be usable in Finnish and the delay is the same 15 seconds, in all attempts. No change in that. Home Assistant Cloud is recognizes Finnish very well, but the delay is still 15 seconds.

**Comment by synesthesiam** [comment]
@HarvsG The VAD in HA core is responsible for detecting the end of speech. However, it is possible for the satellite to do it too and send an empty audio packet to signal the end of the command. This might be worth looking into so we can try silero out in the wild before bringing it into HA core.

**Comment by smoltron** [comment]
I have now a clue of the 15 second behaviour. When I speak from 5 cm distance, the response is immediate. When I am 3 meters away, there is always the 15 second wait, even when the room is totally quiet. The sentences are in any case detected right.
I am using  Jabra 410 hockey puck. It has a very sensitive microphone that can cover a large room easily. I think that the recording threshold must be set right and I have no idea where to start looking for the setting. Is it in local Assist Microphone or in wakeword detection?

**Comment by HarvsG** [comment]
> I have now a clue of the 15 second behaviour. When I speak from 5 cm distance, the response is immediate. When I am 3 meters away, there is always the 15 second wait, even when the room is totally quiet. The sentences are in any case detected right. I am using Jabra 410 hockey puck. It has a very sensitive microphone that can cover a large room easily. I think that the recording threshold must be set right and I have no idea where to start looking for the setting. Is it in local Assist Microphone or in wakeword detection?

Yes I think in part, the speech detection is monitoring the average volume and when the is a "step" down it determines that speech must have ended, so when speech is quiet or background noise high speech end is not detected.

**Comment by smoltron** [comment]
> Yes I think in part, the speech detection is monitoring the average volume and when the is a "step" down it determines that speech must have ended, so when speech is quiet or background noise high speech end is not detected.

In any case, there is a sensitivity problem. The room is quiet and the speech-to-text transformation happens without errors. Jabra 410 has a sensitive microphone and it does not have noise problems. It should be easy to separate signal from low noise.  I do not see any way to do that currently.

**Comment by whitepail** [comment]
It's a problem of webrtc detection of HA. 
To resolve it I'm using silero VAD on satellite to detect end of speech as @synesthesiam adviced.

Also a good idea to move assist_pipeline.AudioSetting -> is_vad_enabled flag as a property of wyoming satellite device and set it in https://github.com/home-assistant/core/tree/dev/homeassistant/components/wyoming/satellite.py accordingly.

On my local HA instance I've set is_vad_enabled for all satellites to False so it won't interfere with SileroVAD.

**Comment by smoltron** [comment]
> It's a problem of webrtc detection of HA. To resolve it I'm using silero VAD on satellite to detect end of speech as @synesthesiam adviced.
> 
> Also a good idea to move assist_pipeline.AudioSetting -> is_vad_enabled flag as a property of wyoming satellite device and set it in https://github.com/home-assistant/core/tree/dev/homeassistant/components/wyoming/satellite.py accordingly.
> 
> On my local HA instance I've set is_vad_enabled for all satellites to False so it won't interfere with SileroVAD.

But how to do it with a local Assist Microphone? Is it necessary to have a  satellite? I cannot locate any setting that could affect.

**Comment by whitepail** [comment]
> But how to do it with a local Assist Microphone? Is it necessary to have a satellite? I cannot locate any setting that could affect.

What do you mean by local Assist? 
Do you use wyoming-satellite for this mic?

**Comment by smoltron** [comment]
> > But how to do it with a local Assist Microphone? Is it necessary to have a satellite? I cannot locate any setting that could affect.
> 
> What do you mean by local Assist? Do you use wyoming-satellite for this mic?

Yes, I understand that this is not the right place. I do not have  a satellite.  The problem is, however, exactly the same with the Assist Microphone addon that is local to home assistant. It is very difficult to find any information or documentation about it, but you can find it from the store.

**Comment by horga83** [comment]
Has there been any progress on this issue?  I experience the same 15 second delay.  I tried the whitepail branch but it crashes, looking through the code it seems to be an approach.  Is there thoughts about adding the pull request or is a different solution proposed?

**Comment by thexperiments** [comment]
Having the same issue on a raspberry pi zero 2w with the 2 mic hat.

Currently running:
```
ExecStart=/home/pi/wyoming-satellite/script/run --name 'Kitchen satellite' \
--uri 'tcp://0.0.0.0:10700' \
--mic-command 'arecord -D plughw:CARD=seeed2micvoicec,DEV=0 -r 16000 -c 1 -f S16_LE -t raw' \
--snd-command 'aplay -D plughw:CARD=seeed2micvoicec,DEV=0 -r 22050 -c 1 -f S16_LE -t raw' \
--mic-auto-gain 5 \
--mic-noise-suppression 2 \
--wake-uri 'tcp://127.0.0.1:10400' \
--wake-word-name 'hey_jarvis' \
--event-uri 'tcp://127.0.0.1:10500'
```


Is there a way to "easily" record what the HA core receives audio wise? I would like to listen to the audio to check if the cause is some interference or wired auto gain behavior.

**Comment by HarvsG** [comment]
> Is there a way to "easily" record what the HA core receives audio wise? I would like to listen to the audio to check if the cause is some interference or wired auto gain behavior.

Yes there is https://www.home-assistant.io/voice_control/troubleshooting/#to-tweak-the-assist-audio-configuration-for-your-device

**Comment by rvsh2** [comment]
Got the same problem here with: satellite openwakeword runing whipser on ha side. Tried every setting possible with wyoming sattelite/openwakeword. Could somebody propose a workaround?

**Comment by horga83** [comment]
I found it really helps to have the whisper model set to tiny-int8.

I have since configured faster-whisper to run remotely on my Arch Linux
desktop machine and it is down to about 2 to 3 seconds with the base-
int8 model which gives somewhat better recognition. 

It still doesn't always solve the end of speech detection and if the
command is not understood I sometimes have to restart faster-whisper as
it hangs.

**Comment by rvsh2** [comment]
I think the solution is to use local VAD as [whitepail](https://github.com/whitepail) suggested. I tried the commited code but it doesn't work in my situation (when wake word is detected it takes 1 sec and the voice activation is going off immediately). Can this solution be merged to the master branch with proper adjustments?

**Comment by horga83** [comment]
Has anyone seen this, curious how it performs:

https://github.com/collabora/WhisperLive

**Comment by olskar** [comment]
I recently assembled a satelite of my own and encounter the same problem. 

I also think the problem is that it does not detect the end of speech. Wakeword is very fast detected, but the yellow leds are on for 15-20 seconds. 

Very happy if this could be solved, thanks!

**Comment by olskar** [comment]
> It's a problem of webrtc detection of HA. To resolve it I'm using silero VAD on satellite to detect end of speech as @synesthesiam adviced.
> 
> Also a good idea to move assist_pipeline.AudioSetting -> is_vad_enabled flag as a property of wyoming satellite device and set it in https://github.com/home-assistant/core/tree/dev/homeassistant/components/wyoming/satellite.py accordingly.
> 
> On my local HA instance I've set is_vad_enabled for all satellites to False so it won't interfere with SileroVAD.

Do you mean to simply do this?

![image](https://github.com/rhasspy/wyoming-satellite/assets/18314560/87130454-6b21-407b-8b82-2526a1cb206a)

**Comment by Vitani** [comment]
Just to share my 2¢, changing the `model` from "medium.en" to "tiny.en" (in Whisper) got rid of the delay, but now it can't understand me, even after increasing the `beam size` from 2 to 5. I tried "small.en" and that had a delay of around half that of "medium.en" (changing `beam size` doesn't appear to make any difference to the delay).

I copied the wav files from my device to my PC to listen to them, and there's just silence at the end, no "wind", no background noise, it seems very reasonable that the recording should have ended (to my human ears at least).

- Device: Rasbperry Pi 3 A+
- Mic/Speaker: Jabra 410 (USB)
- Language: English
- Wake Word: "alexa" openWakeWord (on device)
- Conversation agent: Home Assistant
- Speech-to-text: faster-whisper
- Text-to-speech: piper
